Stack overflow in forkchoice

After an OOM on Witti, I restarted Teku. This was at a point when the chain had over 130 epochs without finalisation.

Resync to head succeeded, but was extremely slow, and produced quite a few of these stack overflow errors in the logs.

I did not specify any memory constraint.

```
2020-06-14 12:34:48.233+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSync Event  *** Current slot: 133374, Head slot: 132600, Connected peers: 4^[[0m
2020-06-14 12:35:00.232+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSync Event  *** Current slot: 133375, Head slot: 132600, Connected peers: 4^[[0m
2020-06-14 12:35:12.232+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSync Event  *** Current slot: 133376, Head slot: 132600, Connected peers: 4^[[0m
2020-06-14 12:35:24.233+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSync Event  *** Current slot: 133377, Head slot: 132600, Connected peers: 6^[[0m
2020-06-14 12:35:36.225+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSync Event  *** Current slot: 133378, Head slot: 132600, Connected peers: 6^[[0m
2020-06-14 12:35:40.287+00:00 | ForkJoinPool.commonPool-worker-7 | ERROR | SyncManager | Error during sync to peer Eth2Peer{id=16Uiu2HAmCGtAV3oz6giNDfVNDfsXidqrhEg2t376nxmwxMLi3yVk, remoteStatus=Optional[PeerStatus{forkDigest=0xf6775d07, finalizedRoot=0xabfa25a62fd102f81bbac4595d7cbc262c3e50fa08710bfea6b78664c156d2c4, finalizedEpoch=4034, headRoot=0x83e60e033445a1caecb782d48199d1e82877477475fb6ca5023562352f365642, headSlot=133373}]}
java.util.concurrent.CompletionException: java.lang.StackOverflowError
        at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1093) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2159) ~[?:?]
        at tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamImpl.completeWithError(ResponseStreamImpl.java:78) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.lambda$cancelRequest$2(Eth2OutgoingRequestHandler.java:168) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.lambda$always$11(SafeFuture.java:198) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.lambda$finish$12(SafeFuture.java:205) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:946) ~[?:?]
        at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2337) ~[?:?]
        at tech.pegasys.teku.util.async.SafeFuture.handle(SafeFuture.java:323) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.finish(SafeFuture.java:202) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.always(SafeFuture.java:198) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.cancelRequest(Eth2OutgoingRequestHandler.java:168) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.cancelRequest(Eth2OutgoingRequestHandler.java:157) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.onAsyncProcessorError(Eth2OutgoingRequestHandler.java:126) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.networking.eth2.rpc.core.AsyncResponseProcessor.lambda$processNextResponse$1(AsyncResponseProcessor.java:114) ~[teku-networking-eth2-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2159) ~[?:?]
        at tech.pegasys.teku.util.async.SafeFuture.lambda$propagateResult$2(SafeFuture.java:95) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2322) ~[?:?]
        at tech.pegasys.teku.util.async.SafeFuture.whenComplete(SafeFuture.java:328) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.whenComplete(SafeFuture.java:26) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.SafeFuture.propagateResult(SafeFuture.java:92) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.util.async.DelayedExecutorAsyncRunner.lambda$runAsync$1(DelayedExecutorAsyncRunner.java:61) ~[teku-util-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1425) [?:?]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) [?:?]
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016) [?:?]
        at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665) [?:?]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598) [?:?]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177) [?:?]
Caused by: java.lang.StackOverflowError
        at tech.pegasys.teku.protoarray.ProtoArrayForkChoiceStrategy.lambda$getProtoNode$2(ProtoArrayForkChoiceStrategy.java:250) ~[teku-protoarray-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.protoarray.ProtoArrayForkChoiceStrategy.getProtoNode(ProtoArrayForkChoiceStrategy.java:248) ~[teku-protoarray-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.protoarray.ProtoArrayForkChoiceStrategy.blockParentRoot(ProtoArrayForkChoiceStrategy.java:240) ~[teku-protoarray-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:94) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]

<snip another few hundred identical lines>
```

A few mins later, this appeared and Teku stopped dead (did not exit, but nothing happening):

```
2020-06-14 12:46:25.371+00:00 | events-14 | ERROR | teku-status-log | PLEASE FIX OR REPORT | Unexpected exception thrown for events-14
java.lang.StackOverflowError: Delayed StackOverflowError due to ReservedStackAccess annotated method
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireShared(AbstractQueuedSynchronizer.java:1323) ~[?:?]
        at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lock(ReentrantReadWriteLock.java:738) ~[?:?]
        at tech.pegasys.teku.protoarray.ProtoArrayForkChoiceStrategy.blockParentRoot(ProtoArrayForkChoiceStrategy.java:238) ~[teku-protoarray-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:94) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
        at tech.pegasys.teku.core.ForkChoiceUtil.get_ancestor(ForkChoiceUtil.java:89) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at tech.pegasys.teku.core.ForkChoiceUtil.lambda$get_ancestor$0(ForkChoiceUtil.java:92) ~[teku-ethereum-core-0.11.4-SNAPSHOT.jar:0.11.4-dev-b3262903]
        at java.util.Optional.flatMap(Optional.java:287) ~[?:?]
<etc - several hundred lines of the same>
```

And on restarting, same thing. Looks like I can't get back on the chain 😢

This is the console output when Teku dies:
```
12:59:23.244 INFO  - Slot Event  *** Slot: 133496, Block:    ... empty, Epoch: 4171, Finalized checkpoint: 4034, Finalized root: abfa25..d2c4, Peers: 0
12:59:31.270 INFO  - Slot Event  *** Slot: 133497, Block:    ... empty, Epoch: 4171, Finalized checkpoint: 4034, Finalized root: abfa25..d2c4, Peers: 0
12:59:40.212 INFO  - Slot Event  *** Slot: 133498, Block:    ... empty, Epoch: 4171, Finalized checkpoint: 4034, Finalized root: abfa25..d2c4, Peers: 0
12:59:52.206 INFO  - Slot Event  *** Slot: 133499, Block:    ... empty, Epoch: 4171, Finalized checkpoint: 4034, Finalized root: abfa25..d2c4, Peers: 5
13:00:00.214 INFO  - Sync Event  *** Current slot: 133500, Head slot: 133428, Connected peers: 6
OpenJDK 64-Bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryAcquireShared(I)I [1]
```

### Versions (Add all that apply)
* Software version: [`teku --version`]
`teku/v0.11.4-dev-b3262903/linux-x86_64/-privatebuild-openjdk64bitservervm-java-13`

--- a/ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java
+++ b/ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java
@@ -84,23 +84,11 @@ public class ForkChoiceUtil {
    */
   private static Optional<Bytes32> get_ancestor(
       ForkChoiceStrategy forkChoiceStrategy, Bytes32 root, UnsignedLong slot) {
-    return forkChoiceStrategy
-        .blockSlot(root)
-        .flatMap(
-            blockSlot -> {
-              if (blockSlot.compareTo(slot) > 0) {
-                return get_ancestor(
-                    forkChoiceStrategy,
-                    forkChoiceStrategy.blockParentRoot(root).orElseThrow(),
-                    slot);
-              } else if (blockSlot.equals(slot)) {
-                return Optional.of(root);
-              } else {
-                // root is older than the queried slot, thus a skip slot. Return earliest root prior
-                // to slot.
-                return Optional.of(root);
-              }
-            });
+    Bytes32 parentRoot = root;
+    while (forkChoiceStrategy.blockSlot(parentRoot).get().compareTo(slot) > 0) {
+      parentRoot = forkChoiceStrategy.blockParentRoot(parentRoot).orElseThrow();
+    }
+    return Optional.of(parentRoot);
   }
 
   /*

--- a/ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java
+++ b/ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java
@@ -85,10 +85,12 @@ public class ForkChoiceUtil {
   private static Optional<Bytes32> get_ancestor(
       ForkChoiceStrategy forkChoiceStrategy, Bytes32 root, UnsignedLong slot) {
     Bytes32 parentRoot = root;
-    while (forkChoiceStrategy.blockSlot(parentRoot).get().compareTo(slot) > 0) {
+    Optional<UnsignedLong> blockSlot = forkChoiceStrategy.blockSlot(root);
+    while (blockSlot.isPresent() && blockSlot.get().compareTo(slot) > 0) {
       parentRoot = forkChoiceStrategy.blockParentRoot(parentRoot).orElseThrow();
+      blockSlot = forkChoiceStrategy.blockSlot(parentRoot);
     }
-    return Optional.of(parentRoot);
+    return blockSlot.isPresent() ? Optional.of(parentRoot) : Optional.empty();
   }
 
   /*

